<a href="https://colab.research.google.com/github/TabithaWKariuki/Fake-News-Detection-Model/blob/main/Fake_News_Detection_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Machine Learning Project

# **Dataset 1** 

## 1. Defining the Analytic Question

To build a model to accurately classify a piece of news as REAL or FAKE using Passive Aggressive algorithms.

## 2. Importing the libraries

In [1]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

## 3. Reading the data

In [2]:
import chardet    
rawdata = open('dataset1.csv', 'rb').read()
result = chardet.detect(rawdata)
charenc = result['encoding']
print(charenc)

Windows-1254


In [3]:
# Reading the data

data = pd.read_csv('dataset1.csv', encoding = 'cp1256')

In [4]:
# Checking the shape of the data

data.shape

(804, 7)

In [ ]:
# Checking the top/first five rows of the data

data.head()

unit_id                                      article_title  \
0  1914947530  Syria attack symptoms consistent with nerve ag...   
1  1914947532  Homs governor says U.S. attack caused deaths b...   
2  1914947533    Death toll from Aleppo bomb attack at least 112   
3  1914947534        Aleppo bomb blast kills six Syrian state TV   
4  1914947535  29 Syria Rebels Dead in Fighting for Key Alepp...   

                                     article_content source       date  \
0  Wed 05 Apr 2017 Syria attack symptoms consiste...    nna   4/5/2017   
1  Fri 07 Apr 2017 at 0914 Homs governor says U.S...    nna   4/7/2017   
2  Sun 16 Apr 2017 Death toll from Aleppo bomb at...    nna  4/16/2017   
3  Wed 19 Apr 2017 Aleppo bomb blast kills six Sy...    nna  4/19/2017   
4  Sun 10 Jul 2016 29 Syria Rebels Dead in Fighti...    nna  7/10/2016   

  location  labels  
0    idlib       0  
1     homs       0  
2   aleppo       0  
3   aleppo       0  
4   aleppo       0

In [ ]:
# Checking the bottom/last five rows of the data

data.tail()

unit_id                                      article_title  \
799  1965511221    Turkish Bombardment Kills 20 Civilians in Syria   
800  1965511222    Martyrs as Terrorists Shell Aleppos Salah Eddin   
801  1965511224  Chemical Attack Kills Five Syrians in Aleppo SANA   
802  1965511226  5 Killed as Russian Military Chopper Shot down...   
803  1965511231  Syrian Army Kills 48 ISIL Terrorists in Deir E...   

                                       article_content source       date  \
799  28-08-2016 Turkish Bombardment Kills 20 Civili...  manar  8/28/2016   
800  17-08-2016 Martyrs as Terrorists Shell Aleppos...  manar   8/1/2016   
801  03-08-2016 Chemical Attack Kills Five Syrians ...  manar   8/3/2016   
802  01-08-2016 5 Killed as Russian Military Choppe...  manar   8/1/2016   
803  April 6 2017 Syrian Army Kills 48 ISIL Terrori...  manar   4/4/2017   

       location  labels  
799      aleppo       1  
800      aleppo       1  
801      aleppo       0  
802       idlib       1  
803  deir ezzor       1

In [ ]:
# Getting the labels/frequencies of the labels column

labels=data.labels
labels.head()

0    0
1    0
2    0
3    0
4    0
Name: labels, dtype: int64

## 4. Modeling using the PassiveAggressive Classifier

In [ ]:
# Spliting the dataset

x_train,x_test,y_train,y_test=train_test_split(data['article_content'], labels, test_size=0.2, random_state=7)

In [ ]:
# Initializing a TfidfVectorizer

tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

# Fit and transform the train set and transform test set

tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

In [ ]:
# Initializing a PassiveAggressiveClassifier

pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

# Predicting on the test set and calculating thr accuracy score of our model
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 47.83%


## 5. Summary

I implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. Accuracy score is 48%, this means the model performed poorly.

# **Dataset 2**

## 6. Reading the data

In [13]:
# Reading the data

data2 = pd.read_csv('dataset2.csv', error_bad_lines=False)

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [14]:
# Checking the shape of the data

data2.shape

(6335, 4)

In [15]:
# Checking the top/first five rows of the data

data2.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [16]:
# Checking the bottom/last five rows of the data

data2.tail()

Unnamed: 0                                              title  \
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332   Anti-Trump Protesters Are Tools of the Oligar...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL

In [18]:
# Getting the labels/frequencies of the label column

label=data2.label
label.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

## 7. Modeling using the PassiveAggressive Classifier

In [20]:
# Spliting the dataset

x_train,x_test,y_train,y_test=train_test_split(data2['text'], label, test_size=0.2, random_state=7)

In [21]:
# Initializing a TfidfVectorizer

tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

# Fit and transform the train set and transform test set

tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

In [22]:
# Initializing a PassiveAggressiveClassifier

pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

# Predicting on the test set and calculating thr accuracy score of our model

y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.9%


## 8. Coclusion

I implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. Accuracy score is 93%, this means the model performed well.